tensorflow/core/kernels/data/reader_dataset_ops.cc

```diff
@@ -378,24 +391,56 @@ class FixedLengthRecordDatasetOp : public DatasetOpKernel {
         mutex_lock l(mu_);
         do {
           // We are currently processing a file, so try to read the next record.
-          if (input_buffer_) {
-            const int64 current_pos = input_buffer_->Tell();
-            DCHECK_GE(file_pos_limit_, 0);
-            if (current_pos < file_pos_limit_) {
+          if (buffered_input_stream_) {
+            const int64 current_pos = buffered_input_stream_->Tell();
+            if (dataset()->compression_type_ == "") {
```

Comment:
The logic in here might be clearer if you divided `Iterator` into two classes (`UncompressedIterator` and `GzipCompressedIterator`) that each contain the relevant implementations for `GetNextInternal()` and `RestoreInternal()`.